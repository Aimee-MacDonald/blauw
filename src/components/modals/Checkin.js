import React, {useContext} from 'react'
import {connect} from 'react-redux'

import {ServerConnectionContext} from '../../util/ServerConnection'

import {setModal} from '../../state_management/actions/modal'
import {checkIn} from '../../state_management/actions/bookings'
import {getSelectedBooking} from '../../state_management/selectors/bookings'

export const Checkin = props => {
  const connection = useContext(ServerConnectionContext)

  const checkBookingIn = () => {
    props.dispatch(checkIn(props.booking._id))
    connection.dispatch({reducer: 'bookings', action: checkIn(props.booking._id)})
    props.dispatch(setModal())
  }

  return(
    <div>
      <h1>Check this booking in?</h1>
      <button onClick={checkBookingIn}>Yes</button>
      <button onClick={() => props.dispatch(setModal())}>No</button>
    </div>
  )
}

const mapStateToProps = ({bookings}) => ({
  booking: getSelectedBooking(bookings)
})

export default connect(mapStateToProps)(Checkin)