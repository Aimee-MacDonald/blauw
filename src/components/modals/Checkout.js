import React, {useContext} from 'react'
import {connect} from 'react-redux'

import {ServerConnectionContext} from '../../util/ServerConnection'

import {checkOut} from '../../state_management/actions/bookings'
import {setModal} from '../../state_management/actions/modal'
import {getSelectedBooking} from '../../state_management/selectors/bookings'

export const Checkout = props => {
  const server = useContext(ServerConnectionContext)

  const checkBookingOut = bookingId => {
    props.dispatch(checkOut(bookingId))
    server.dispatch({reducer: 'bookings', action: checkOut(bookingId)})
    props.dispatch(setModal())
  }

  return(
    <div>
      <h1>Check this booking out?</h1>
      <button onClick={() => checkBookingOut(props.booking._id)}>Yes</button>
      <button onClick={() => props.dispatch(setModal())}>No</button>
    </div>
  )
}

const mapStateToProps = ({bookings}) => ({booking: getSelectedBooking(bookings)})

export default connect(mapStateToProps)(Checkout)