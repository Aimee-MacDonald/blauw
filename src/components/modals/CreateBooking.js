import React, {useContext} from 'react'
import styled from 'styled-components'
import {connect} from 'react-redux'

import {ServerConnectionContext} from '../../util/ServerConnection'

import {setModal} from '../../state_management/actions/modal'
import {createBooking} from '../../state_management/actions/bookings'

const CreateBooking = props => {
  const connection = useContext(ServerConnectionContext)

  const saveBooking = e => {
    e.preventDefault()

    const bookingDetails = {
      name: e.target.name.value,
      date: parseInt(e.target.date.value),
      room: parseInt(e.target.room.value),
      nights: parseInt(e.target.nights.value)
    }

    const createBookingAction = createBooking(bookingDetails)

    props.dispatch(createBookingAction)
    connection.dispatch({reducer: 'bookings', action: createBookingAction})
    props.dispatch(setModal())
  }

  return(
    <StyledCreateBooking onSubmit={saveBooking}>
      <h1>Create Booking</h1>

      <label htmlFor='name'>Name:</label>
      <input id='name' />

      <label htmlFor='date'>Date:</label>
      <input id='date' />

      <label htmlFor='room'>Room:</label>
      <input id='room' />

      <label htmlFor='nights'>Nights:</label>
      <input id='nights' />
      
      <button type='reset' onClick={() => props.dispatch(setModal())}>Cancel</button>
      <button type='submit'>Save</button>
    </StyledCreateBooking>
  )
}

const StyledCreateBooking = styled.form`
  background-color: lime;
`

export default connect()(CreateBooking)