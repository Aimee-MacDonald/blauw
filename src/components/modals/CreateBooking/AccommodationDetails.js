import React from 'react'
import styled from 'styled-components'
import {connect} from 'react-redux'

import {getRoomDetails, getRoomOptions} from '../../../state_management/selectors/rooms'

export const AccommodationDetails = props => {
  const setGuestName = (roomIndex, guestIndex, guestName) => {
    props.editGuestDetails(roomIndex, guestIndex, {
      ...props.rooms[roomIndex].guests[guestIndex],
      name: guestName
    })
  }


  const setGuestArrival = (roomIndex, guestIndex, guestArrival) => {
    props.editGuestDetails(roomIndex, guestIndex, {
      ...props.rooms[roomIndex].guests[guestIndex],
      arrival: guestArrival
    })
  }

  const setGuestDeparture = (roomIndex, guestIndex, guestDeparture) => {
    props.editGuestDetails(roomIndex, guestIndex, {
      ...props.rooms[roomIndex].guests[guestIndex],
      departure: guestDeparture
    })
  }

  return(
    <StyledAccommodationDetails>
      <p>Room</p>
      <p>Guests</p>
      <p>Arrival</p>
      <p>Departure</p>
      <p>Nights</p>
      <p>Price</p>

      {props.rooms && props.rooms.map(room => (
        <StyledRoomDetails key={room.index}>
          <select onChange={e => props.selectRoomOption(e.target.value)}>
            <option>Select Room</option>
            {getRoomOptions(props.stateRooms).map(roomOption => (
              <option key={roomOption._id} value={roomOption._id}>{roomOption.name}</option>
            ))}
          </select>

          {room.guests.map((guest, index) => (
            <StyledGuestDetails key={`${guest}_${index}`}>
              <input
                value={guest.name}
                onChange={e => setGuestName(room.index, index, e.target.value)}
              />

              <input
                type='date'
                value={guest.arrival}
                onChange={e => setGuestArrival(room.index, index, e.target.value)}
              />

              <input
                type='date'
                value={guest.departure}
                onChange={e => setGuestDeparture(room.index, index, e.target.value)}
              />

              <p>{guest.nights}</p>
              <p>{`R ${guest.price}`}</p>
            </StyledGuestDetails>
          ))}
        </StyledRoomDetails>
      ))}

      <div id='addButtons'>
        <button type='button' onClick={props.addRoom}>Add Room</button>
        {props.selectedRoom &&
          !!props.rooms[props.selectedRoom - 1].roomId &&
          props.rooms[props.selectedRoom - 1].guests.length < getRoomDetails(props.rooms[props.selectedRoom - 1].roomId, props.stateRooms).maxPax &&
          <button type='button' onClick={props.addGuest}>Add Guest</button>
        }
      </div>

      <p id='total'>Total: R</p>
    </StyledAccommodationDetails>
  )
}

const StyledAccommodationDetails = styled.div`
  grid-column-start: 1;
  grid-column-end: 5;
  display: grid;
  grid-template-columns: repeat(6, 1fr);
  width 64rem;
  text-align: center;

  #addButtons{
    grid-column-start: 1;
    grid-column-end: 3;
    display: flex;
    justify-content: space-evenly;
  }

  #total{
    grid-column-start: 6;
  }
`

const StyledRoomDetails = styled.div`
  grid-column-start: 1;
  grid-column-end: 7;
  display: grid;
  grid-template-columns: repeat(6, 1fr);
`

const StyledGuestDetails = styled.div`
  grid-column-start: 2;
  grid-column-end: 7;
  display: grid;
  grid-template-columns: repeat(5, 1fr);
`

const mapStateToProps = ({rooms}) => ({stateRooms: rooms})

export default connect(mapStateToProps)(AccommodationDetails)