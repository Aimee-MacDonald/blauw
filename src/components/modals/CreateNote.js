import React from 'react'
import styled from 'styled-components'
import {connect} from 'react-redux'

import {setModal} from '../../state_management/actions/modal'
import {createNote} from '../../state_management/actions/notes'

export const CreateNote = props => {
  const addNote = e => {
    e.preventDefault()

    const note = {
      title: e.target.title.value,
      content: e.target.content.value
    }

    props.dispatch(createNote(note))
    props.dispatch(setModal())
  }

  return(
    <StyledCreateNote onSubmit={addNote}>
      <h1>Create Note</h1>

      <label htmlFor='title'>Title</label>
      <input id='title'/>

      <textarea id='content'/>

      <button type='button' onClick={() => props.dispatch(setModal())}>Cancel</button>
      <button type='submit'>Save</button>
    </StyledCreateNote>
  )
}

const StyledCreateNote = styled.form`
  display: grid;
  grid-template-columns: 1fr 1fr;

  h1, textarea{
    grid-column-start: 1;
    grid-column-end: 3;
  }
`

export default connect()(CreateNote)