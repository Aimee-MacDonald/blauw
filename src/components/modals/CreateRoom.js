import React, {useContext} from 'react'
import {connect} from 'react-redux'
import styled from 'styled-components'

import {ServerConnectionContext} from '../../util/ServerConnection'
import {createRoom} from '../../state_management/actions/rooms'

import {setModal} from '../../state_management/actions/modal'

export const CreateRoom = props => {
  const server = useContext(ServerConnectionContext)

  const saveRoom = e => {
    e.preventDefault()

    const createRoomAction = createRoom({
      name: e.target.name.value,
      group: e.target.group.value,
      shared: e.target.shared.checked,
      maxPax: e.target.maxPax.value,
      basePrice: e.target.basePrice.value,
      personPrice: e.target.personPrice.value,
      showBeds: e.target.showBeds.checked
    })

    props.dispatch(createRoomAction)
    server.dispatch({reducer: 'rooms', action: createRoomAction})
    props.dispatch(setModal())
  }

  return(
    <StyledCreateRoom onSubmit={saveRoom}>
      <h1>Add Room</h1>

      <label htmlFor='name'>Name</label>
      <input id='name'/>

      <label htmlFor='group'>Group</label>
      <input id='group'/>

      <label htmlFor='shared'>Shared</label>
      <input id='shared' type='checkbox'/>

      <label htmlFor='maxPax'>Max Pax</label>
      <input id='maxPax'/>

      <label htmlFor='basePrice'>Base Price</label>
      <input id='basePrice'/>

      <label htmlFor='personPrice'>Person Price</label>
      <input id='personPrice'/>

      <label htmlFor='showBeds'>Show Beds</label>
      <input id='showBeds' type='checkbox'/>

      <button type='reset' onClick={() => props.dispatch(setModal())}>Cancel</button>
      <button type='submit'>Save</button>
    </StyledCreateRoom>
  )
}

const StyledCreateRoom = styled.form`
  display: grid;
  grid-template-columns: 1fr 1fr;
  padding: 1.6rem;

  h1{
    grid-column-start: 1;
    grid-column-end: 3;
    margin-bottom: 3.2rem;
    text-align: center;
  }

  label{
    text-align: right;
    margin-top: 0.4rem;
    padding-right: 0.8rem;
  }

  input{
    background-color: transparent;
    border: none;
    border-bottom 1px solid black;
    width: 75%;
    margin-top: 0.4rem;
  }

  button{
    padding: 0.4rem;
    margin: 3.2rem 0.1rem 0 0.2rem;
  }
`

export default connect()(CreateRoom)