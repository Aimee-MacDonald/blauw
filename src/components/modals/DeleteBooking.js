import React, {useContext} from 'react'
import {connect} from 'react-redux'

import {ServerConnectionContext} from '../../util/ServerConnection'

import {deleteBooking} from '../../state_management/actions/bookings'
import {setModal} from '../../state_management/actions/modal'

const DeleteBooking = props => {
  const connection = useContext(ServerConnectionContext)

  return(
    <div>
      <h1>Delete this booking?</h1>
      <p>{props._id}</p>
      <button onClick={() => {
        props.dispatch(deleteBooking(props._id))
        connection.dispatch({reducer: 'bookings', action: deleteBooking(props._id)})
        props.dispatch(setModal())
      }}>Yes</button>
      <button onClick={() => props.dispatch(setModal())}>No</button>
    </div>
  )
}

export default connect()(DeleteBooking)