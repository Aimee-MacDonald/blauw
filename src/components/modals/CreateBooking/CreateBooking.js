import React from 'react'
import styled from 'styled-components'

import AccommodationDetails from './AccommodationDetails'

export const CreateBooking = () => (
  <StyledCreateBooking>
    <h1>Create Booking</h1>

    <div id='labeled_booking_name'>
      <label htmlFor='booking_name'>Booking Name</label>
      <input id='booking_name'/>
    </div>

    <AccommodationDetails/>

    <div id='labeled_notes'>
      <label htmlFor='notes'>Notes</label>
      <textarea/>
    </div>

    <div id='formControls'>
      <button>Save</button>
      <button>Cancel</button>
    </div>
  </StyledCreateBooking>
)

const StyledCreateBooking = styled.form`
  display: grid;
  grid-template-columns: repeat(4, 1fr);

  h1{
    grid-column-start: 1;
    grid-column-end: 3;
  }

  #labeled_booking_name{
    grid-column-start: 1;
    grid-column-end: 5;
    display: flex;
  }

  #labeled_notes{
    grid-column-start: 1;
    grid-column-end: 5;
    display: flex;
    flex-flow: nowrap column;

    textarea{
      width: 100%;
    }
  }

  #formControls{
    grid-column-start: 2;
    grid-column-end: 4;
    display: flex;
    flex-flow: nowrap row;
    justify-content: space-evenly;
  }
`

export default CreateBooking