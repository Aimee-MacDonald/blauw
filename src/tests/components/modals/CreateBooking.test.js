import React from 'react'
import {shallow} from 'enzyme'

import {CreateBooking} from '../../../components/modals/CreateBooking'

test(`Render default 'Create Booking' modal`, () => {
  const rooms = [{
    roomName: 'camping',
    groupName: 'Camping',
    maxPax: 80
  }, {
    roomName: 'safari_tent_1',
    groupName: 'Safari Tents',
    maxPax: 2
  }, {
    roomName: 'safari_tent_2',
    groupName: 'Safari Tents',
    maxPax: 2
  }]

  const wrapper = shallow(<CreateBooking
    rooms={rooms}
  />)

  expect(wrapper.getElement()).toMatchSnapshot()
})