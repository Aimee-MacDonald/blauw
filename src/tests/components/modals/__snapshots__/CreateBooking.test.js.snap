// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render default 'Create Booking' modal 1`] = `
<styled.form>
  <h1>
    Create Booking
  </h1>
  <div
    id="labeled_booking_name"
  >
    <label
      htmlFor="booking_name"
    >
      Booking name:
    </label>
    <input
      id="booking_name"
    />
  </div>
  <div
    id="accommodation_details"
  >
    <div>
      <p>
        Room
      </p>
      <p>
        Guest
      </p>
      <p>
        Arrival
      </p>
      <p>
        Departure
      </p>
      <p>
        Nights
      </p>
      <p>
        Price
      </p>
    </div>
    <div>
      <input />
      <section>
        <input />
        <input
          type="date"
        />
        <input
          type="date"
        />
        <input />
        <p>
          Price
        </p>
      </section>
    </div>
  </div>
  <button>
    Add Room
  </button>
  <button>
    Add Guest
  </button>
  <p
    id="total"
  >
    Total: R
  </p>
  <textarea />
  <div
    id="form_controls"
  >
    <button>
      Create Booking
    </button>
    <button>
      Cancel
    </button>
  </div>
</styled.form>
`;
